pkg/controllers/deprovisioning/drift.go

```diff
@@ -0,0 +1,103 @@
+/*
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package deprovisioning
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	v1 "k8s.io/api/core/v1"
+	"knative.dev/pkg/logging"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/aws/karpenter-core/pkg/apis/config/settings"
+	"github.com/aws/karpenter-core/pkg/apis/v1alpha5"
+	"github.com/aws/karpenter-core/pkg/controllers/provisioning"
+	"github.com/aws/karpenter-core/pkg/controllers/state"
+	"github.com/aws/karpenter-core/pkg/metrics"
+)
+
+// Drift is a subreconciler that deletes empty nodes.
+// Drift will respect TTLSecondsAfterEmpty
+type Drift struct {
+	kubeClient  client.Client
+	cluster     *state.Cluster
+	provisioner *provisioning.Provisioner
+}
+
+func NewDrift(kubeClient client.Client, cluster *state.Cluster, provisioner *provisioning.Provisioner) *Drift {
+	return &Drift{
+		kubeClient:  kubeClient,
+		cluster:     cluster,
+		provisioner: provisioner,
+	}
+}
+
+// ShouldDeprovision is a predicate used to filter deprovisionable nodes
+func (d *Drift) ShouldDeprovision(ctx context.Context, n *state.Node, provisioner *v1alpha5.Provisioner, nodePods []*v1.Pod) bool {
+	// Look up the feature flag to see if we should deprovision the node because of drift.
+	if !settings.FromContext(ctx).DriftEnabled {
+		return false
+	}
+	return n.Node.Annotations[v1alpha5.VoluntaryDisruptionAnnotationKey] == v1alpha5.VoluntaryDisruptionDriftedAnnotationValue
+}
+
+// ComputeCommand generates a deprovisioning command given deprovisionable nodes
+func (d *Drift) ComputeCommand(ctx context.Context, candidates ...CandidateNode) (Command, error) {
+	pdbs, err := NewPDBLimits(ctx, d.kubeClient)
```

Comment:
musing: Some of these things are starting to look repetitive and have a common thread among the other deprovisioners. I'm wondering if we can simplify this and abstract some of these common checks out of this call